### PR TITLE
Add `history delete` #4929

### PR DESCRIPTION
**Pull request by HackTheOxidation**
This MR adds the `delete` command to `history`, which allows the user to only delete commands from the history,
that matches a specific pattern (in the form of a regular expression).

This is done by extending `xonsh.history.base.History` and `xonsh.history.main.HistoryAlias` as well as provide default implementations for the `JsonHistory` and `SqliteHistory` backends.

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

File: xonsh/history/sqlite.py [hunk 1]
```diff
@@ -220,6 +221,15 @@ def xh_sqlite_wipe_session(sessionid=None, filename=None):
         c.execute(sql, (str(sessionid),))
 
 
+def xh_sqlite_delete_input_matching(pattern, filename=None):
+    """Deletes entries from the database where the input matches a pattern."""
+    sql = f"DELETE FROM xonsh_history WHERE inp LIKE '{pattern}' ESCAPE '\\';"
```
Comment (jnoortheen) [inline xonsh/history/sqlite.py:226]: Here the like pattern is different from regex pattern right?

Comment (HackTheOxidation) [reply xonsh/history/sqlite.py:226]: Yes, I should probably rename it to `sql_pattern` or something similar to make it clear?

Comment (jnoortheen) [reply xonsh/history/sqlite.py:226]: what I meant was that the json implementation and this is not the same. `LIKE` operator doesn't implement the regex

Comment (HackTheOxidation) [reply xonsh/history/sqlite.py]: Ah, I think I understand :sweat_smile: 

When reading the sqlite documentation for [REGEXP](https://www.sqlite.org/lang_expr.html),
it says that there is no `regexp()`-function defined by default.

Is there anything in the `xonsh` code base that defines this, or is it fine to just assume that it is defined? :thinking:

Comment (jnoortheen) [reply xonsh/history/sqlite.py]: The base function can iterate over the entries and delete those matching the pattern. The sub-classes can implement an option to delete an item at the given index

**Review by jnoortheen — APPROVED** [review]

File: xonsh/history/sqlite.py [hunk 2]
```diff
@@ -220,6 +221,17 @@ def xh_sqlite_wipe_session(sessionid=None, filename=None):
...
 
+def xh_sqlite_delete_input_matching(pattern, filename=None):
+    """Deletes entries from the database where the input matches a pattern."""
+    with _xh_sqlite_get_conn(filename=filename) as conn:
+        c = conn.cursor()
+        _xh_sqlite_create_history_table(c)
+        for inp, *_ in _xh_sqlite_get_records(c):
+            if pattern.match(inp):
+                sql = f"DELETE FROM xonsh_history WHERE inp LIKE '{inp}'"
+                c.execute(sql)
```
Comment (jnoortheen) [inline xonsh/history/sqlite.py:232]: Instead of using like you can use = sign. If the input itself had '%' in it , it might not properly escape it.

Comment (HackTheOxidation) [reply xonsh/history/sqlite.py:232]: Great! Thanks. I've changed the code to use the '='-operator instead. :+1:

**Comment by jnoortheen** [comment]
LGTM, Thanks @HackTheOxidation !

**Pull request merged**